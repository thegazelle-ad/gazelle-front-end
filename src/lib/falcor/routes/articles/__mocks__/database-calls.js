export async function getPaginatedArticle(pageLength, pageIndex) {
  const articles = [];
  for (
    let i = pageIndex * pageLength;
    i < pageIndex * pageLength + pageLength;
    i++
  ) {
    articles.push({ slug: `slug-${i}` });
  }
  return articles;
}
