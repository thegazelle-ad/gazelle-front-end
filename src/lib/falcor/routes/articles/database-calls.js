/**
 * @typedef PaginationArticle
 * @param {string} slug - The slug of the article
 */

/**
 * Fetches a page of articles where pages are a given length
 * @param {any} database - The knex instance to query on
 * @param {number} pageLength - Length of page to be fetched
 * @param {number} pageIndex - Which page to fetch of size pageLength
 * @returns {Promise<PaginationArticle[]>} - An array of articles on the page
 */
export async function getPaginatedArticle(database, pageLength, pageIndex) {
  const offset = pageLength * pageIndex;
  const articles = await database
    .select('slug')
    .from('articles')
    .orderBy('created_at', 'DESC')
    .limit(pageLength)
    .offset(offset);
  return articles;
}
