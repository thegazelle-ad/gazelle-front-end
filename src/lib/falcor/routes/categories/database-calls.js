/**
 * Fetches direct meta data of categories from the categories database table
 * @param {any} database - The Knex object for the databse to update
 * @param {string} queryField - Indicates which field to query by
 * @param {string[]} queryParams - Array of parameters of type queryField of categories to fetch
 * @param {string[]} columns - Which columns of the articles table to fetch
 * @returns {Promise<Object[]>}
 */
export function categoryQuery(database, queryField, queryParams, columns) {
  // In order to be able to identify the rows we get back we need to include the queryField
  if (!columns.includes(queryField)) {
    columns.push(queryField);
  }
  return database
    .select(...columns)
    .from('categories')
    .whereIn(queryField, queryParams);
}
