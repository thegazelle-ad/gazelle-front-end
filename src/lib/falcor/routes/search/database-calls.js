import { database } from 'lib/db';

export async function searchTagsQuery(queries, min, max) {
  const results = {};

  const dbUpdate = queries.map(query =>
    database
      .select('slug')
      .from('tags')
      .where('name', 'like', `%${query}%`)
      .limit(max - min + 1)
      .offset(min),
  );

  const rowsArray = await Promise.all(dbUpdate);
  queries.map((query, index) => {
    results[query] = rowsArray[index].map(row => row.slug);
    return true;
  });
  return results;
}
