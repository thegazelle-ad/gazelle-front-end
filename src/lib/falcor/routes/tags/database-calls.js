import _ from 'lodash';

import { database } from 'lib/db';

/**
 * @param   {Object[]}   tagObject the tag to be added.
 * @returns {boolean}    tags      number of tags modified
 */
export async function addTag(tagObject) {
  await database('tags').insert(tagObject);
  return true;
}

/**
 * Queries the tags table in the database for all tags with the given slugs selecting for
 * the given columns.
 * @param  {string[]}    slugs   the tags to select
 * @param  {string[]}    columns the columns of tags to return
 * @returns{Object[]}    rows    the result.
 */
export async function tagQuery(slugs, columns) {
  if (!_.isArray(slugs) || !_.isArray(columns)) {
    throw new Error(
      `tagQuery must be called with slugs and columns as arrays. Received slugs ${slugs} and columns ${columns}`,
    );
  }

  const processedColumns = _.uniq(columns.concat(['slug']));
  const rows = await database
    .select(...processedColumns)
    .from('tags')
    .whereIn('slug', slugs)
    .catch(e => {
      throw new Error(e);
    });
  return rows;
}

/**
 * Update tags given a json graph.
 * @param   {Object}   jsonGraphArg - the json graph to update
 */
export function updateTags(jsonGraphArg) {
  _.forEach(jsonGraphArg, async (tagObject, slug) => {
    await database('tags')
      .where('slug', '=', slug)
      .update(tagObject)
      .then(data => {
        if (data !== 1) {
          throw new Error(`Problems updating tag with slug: ${slug}`);
        }
      })
      .catch(e => {
        throw e;
      });
  });
}
