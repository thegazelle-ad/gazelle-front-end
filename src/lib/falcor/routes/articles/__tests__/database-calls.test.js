import { getPaginatedArticle } from '../database-calls';
import {
  getDatabaseConnection,
  initializeTestDatabase,
  cleanupTestDatabase,
} from '__tests__/helpers/db-helpers';
import path from 'path';

const name = path
  .basename(__filename)
  .replace(/[.-]/g, '_')
  .replace(/^_+/, '');

const databaseName = `gazelle_test_${name}_${new Date().getTime()}`;

const database = getDatabaseConnection(databaseName);

const maxSlugNum = 200;

// End the knex connection after all tests are done
afterAll(() => database.destroy());

describe('getPaginatedArticle', () => {
  // We can use All instead of Each as these are only select queries
  beforeAll(() => initializeTestDatabase(database, databaseName));
  afterAll(() => cleanupTestDatabase(databaseName));

  it('fetches expected rows from database', async () => {
    await expect(
      getPaginatedArticle(database, 5, 0),
    ).resolves.toMatchSnapshot();
  });

  it('fetches the newest article for page 0', async () => {
    await expect(getPaginatedArticle(database, 1, 0)).resolves.toEqual([
      { slug: `slug-${maxSlugNum}` },
    ]);
  });

  it('correctly computes offset', async () => {
    await expect(getPaginatedArticle(database, 1, 5)).resolves.toEqual([
      { slug: `slug-${maxSlugNum - 5}` },
    ]);
  });

  it('throws on negative page index', async () => {
    await expect(getPaginatedArticle(database, 1, -1)).rejects.toThrow();
  });

  it('throws on negative page length', async () => {
    await expect(getPaginatedArticle(database, -1, 5)).rejects.toThrow();
  });

  it('returns nothing on zero page length', async () => {
    await expect(getPaginatedArticle(database, 0, 5)).resolves.toEqual([]);
  });
});
