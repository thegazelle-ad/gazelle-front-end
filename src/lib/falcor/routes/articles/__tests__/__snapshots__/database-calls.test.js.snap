// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPaginatedArticle fetches expected rows from database 1`] = `
Array [
  Object {
    "slug": "slug-200",
  },
  Object {
    "slug": "slug-199",
  },
  Object {
    "slug": "slug-198",
  },
  Object {
    "slug": "slug-197",
  },
  Object {
    "slug": "slug-196",
  },
]
`;
